src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -712,17 +712,26 @@ public void onLoad(FlowExecutionOwner owner) throws IOException {
         this.owner = owner;
 
         try {
+            initializeStorage();  // Throws exception and bombs out if we can't load FlowNodes
+        } catch (Exception ex) {
             try {
-                initializeStorage();  // Throws exception and bombs out if we can't load FlowNodes
-            } catch (Exception ex) {
-                LOGGER.log(Level.WARNING, "Error initializing storage and loading nodes, will try to create placeholders for: "+this, ex);
+                if (!canResume()) {
+                    // This case is expected for PERFORMANCE_OPTIMIZED Pipelines after a Jenkins crash (JENKINS-55287). The specific issue is
+                    // usually that only the FlowStartNode has been persisted and so we cannot find the current head node in storage.
+                    String message = "Unable to resume " + owner.getExecutable() + " because it was not saved before Jenkins shut down. Did Jenkins crash?";
+                    LOGGER.log(Level.WARNING, message, ex);
+                    ex = new AbortException(message); // Change the exception displayed in the build log to one that is more useful for users.
+                } else {
+                    LOGGER.log(Level.WARNING, "Unable to resume " + owner.getExecutable() + " because its flow nodes could not be loaded", ex);
+                }
                 createPlaceholderNodes(ex);
-                return;
+            } catch (Exception ex2) {
```

Comment:
This is all mostly refactoring. The old code had this structure:
```
try {
  try {
    // code that can throw
  } catch (first) {
    // code that can throw
    return
  }
} catch (second) {
   throw second
}
```

Now it has this structure:

```
try {
  // code that can throw
} catch (first) {
  try {
    // code that can throw
  } catch (second) {
    second.addSuppressed(first)
    throw second
  }
  return
}
```